wasm/src/utils.rs

```diff
@@ -1,16 +1,115 @@
 use petgraph::visit::{EdgeRef, IntoEdgeReferences, IntoNodeReferences};
 use std::{collections::VecDeque, error::Error, fmt};
 use wasm_bindgen::prelude::*;
+use web_time::Instant;
 
+#[cfg(not(feature = "test"))]
+#[wasm_bindgen(module = "src/logger/index.ts")]
+extern "C" {
+    #[wasm_bindgen(js_name = log)]
+    pub static LOGGER: Logger;
+
+    pub type Logger;
+    #[wasm_bindgen(method)]
+    pub fn debug(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn info(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn warn(this: &Logger, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn error(this: &Logger, message: &str);
+}
+
+#[cfg(feature = "test")]
 #[wasm_bindgen]
 extern "C" {
-    fn alert(s: &str);
-    // Use `js_namespace` here to bind `console.log(..)` instead of just
-    // `log(..)`
-    #[wasm_bindgen(js_namespace = console, js_name = log)]
-    pub fn log_str(s: &str);
-    #[wasm_bindgen(js_namespace = console)]
-    pub fn log(s: String);
+    #[wasm_bindgen(js_name = console)]
+    pub static LOGGER: Console;
+
+    pub type Console;
+    #[wasm_bindgen(method)]
+    pub fn debug(this: &Console, message: &str);
+    #[wasm_bindgen(method, js_name = log)]
+    pub fn info(this: &Console, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn warn(this: &Console, message: &str);
+    #[wasm_bindgen(method)]
+    pub fn error(this: &Console, message: &str);
+}
+
+pub struct PerfLogger {
+    name: String,
+    start: Instant,
+    elapsed: Option<u128>,
+    splits: Vec<PerfLogger>,
+}
+
+impl PerfLogger {
+    pub fn new(name: String) -> Self {
+        PerfLogger {
+            name,
+            start: Instant::now(),
+            splits: Vec::new(),
+            elapsed: None,
+        }
+    }
+
+    pub fn start_split(&mut self, name: String) -> &mut PerfLogger {
+        // stop the last split
+        self.stop_split();
+
+        // create a new split
+        self.splits.push(PerfLogger::new(name));
+        self.splits.last_mut().unwrap()
+    }
+
+    pub fn stop_split(&mut self) {
+        self.splits.last_mut().map(|split| {
+            if !split.stopped() {
+                split.stop()
+            }
+        });
+    }
+
+    pub fn stop(&mut self) {
+        if self.stopped() {
+            LOGGER.warn(&format!("PerfLogger {} is already stopped", self.name));
+        } else {
```

Comment:
If you double-check that it's not already stopped in this function, it feels unecessary to check this before _calling_ `self.stop()` elsewhere

Reply:
Kind of, the check here is to make sure that the place that is calling this checked stopped before.